express-notes-api/index.js

```diff
@@ -0,0 +1,173 @@
+const express = require('express');
+const app = express();
+const data = require('./data.json');
+const fs = require('fs');
+const JSONMiddleware = express.json();
+app.use(JSONMiddleware);
+
+// require automatically parses
+
+app.get('/api/notes', function (req, res) {
+  const array = [];
+  for (const property in data.notes) {
+    array.push(data.notes[property]);
+  }
+  res.status(200);
+  res.send(array);
+});
+
+app.get('/api/notes/:id', function (req, res) {
+  const reqId = Number(req.params.id);
+  if (isNaN(reqId)) {
+    const errorObject =
+    {
+      error: 'id must be a positive integer'
+    };
+    res.status(400);
+    res.send(errorObject);
+    return;
+  }
+  const array = [];
+
+  if (!data.notes[reqId]) {
+    const cannotFindObject =
+    {
+      error: `cannot find note with id ${reqId}`
+    };
+    res.status(404);
+    res.send(cannotFindObject);
+    return;
+  } else {
+    array.push(data.notes[reqId]);
+  }
+
+  res.send(array[0]);
+
+});
+
+let nextId = data.nextId;
+
+app.post('/api/notes', function (req, res) {
+  const reqBody = req.body;
+  if (!reqBody.content) {
+    const errorObject =
+    {
+      error: 'content is a required field'
+    };
+    res.status(400);
+    res.send(errorObject);
+  } else if (reqBody.content) {
+    reqBody.id = nextId;
+    data.notes[nextId] = reqBody;
+    fs.writeFile('data.json', JSON.stringify(data, null, 2), 'utf-8', function (err) {
+      if (err) {
+        console.error(err);
+        const fiveHundredObject =
+        {
+          error: 'An unexpected error occured'
+        };
+        res.status(500);
+        res.send(fiveHundredObject);
+      }
+      nextId++;
+      res.status(201);
+      res.send(reqBody);
+    });
+  }
+});
+
+app.delete('/api/notes/:id', (req, res) => {
+  const reqId = Number(req.params.id);
+
+  if (isNaN(reqId)) {
+    const fourHundrerErrorObject =
+    {
+      error: 'Id must be a positive integer'
+    };
+    res.status(400);
+    res.send(fourHundrerErrorObject);
+  }
+
+  if (!data.notes[reqId]) {
+    const cannotFindObject = {
+      error: `cannot find note with id ${reqId}`
+    };
+    res.status(404);
+    res.send(cannotFindObject);
+
+  } else {
+    delete data.notes[reqId];
+    fs.writeFile('data.json', JSON.stringify(data, null, 2), 'utf-8', function (err) {
+      if (err) {
+        console.error(err);
+        const fiveHundredObject =
+        {
+          error: 'An unexpected error occured'
+        };
+        res.status(500);
+        res.send(fiveHundredObject);
+      }
+      res.sendStatus(204);
```

Comment:
Ok, great, the only tiny change needed is to include a `return` statement at the bottom of your `if` blocks so that your server doesn't try to respond with a 500 and then a 204 immediately after. Check your other routes for the same situation.